### PR TITLE
Add optional channel arguments.

### PR DESCRIPTION
**Pull request by fabricematrat**

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/theblues-lib/60/
Test PASSed.

File: theblues/charmstore.py [hunk 1]
```diff
@@ -70,23 +70,32 @@ def _get(self, url):
...
                               exc.message)
 
-    def _meta(self, entity_id, includes):
+    def _meta(self, entity_id, includes, channel=None):
         '''Retrieve metadata about an entity in the charmstore.
 
         @param entity_id The ID either a reference or a string of the entity
                to get.
         @param includes Which metadata fields to include in the response.
+        @param channel optional channel information.
```
Comment (frankban) [inline theblues/charmstore.py]: optional channel name?

File: theblues/charmstore.py [hunk 2]
```diff
@@ -118,67 +127,90 @@ def entities(self, entity_ids):
...
         @param filename The name of the file in the archive.
+        @param channel optional channel information.
         '''
-        return '%s/%s/archive/%s' % (self.url, _get_path(entity_id), filename)
-
-    def files(self, entity_id, manifest=None, filename=None, read_file=False):
+        url = '{}/{}/archive/{}'.format(self.url, _get_path(entity_id),
+                                        filename)
+        if channel is not None:
+            url = '{}?channel={}'.format(url, channel)
```
Comment (frankban) [inline theblues/charmstore.py]: This logic is repeated multiple times. We should have it defined in a single place, and I like the urlencode approach you use above.

File: theblues/charmstore.py [hunk 3]
```diff
@@ -198,16 +230,20 @@ def files(self, entity_id, manifest=None, filename=None, read_file=False):
         @param filename The name of the file in the archive to get.
         @param read_file Whether to get the url for the file or the file
         contents.
+        @param channel optional channel information.
         '''
         if manifest is None:
-            manifest_url = '%s/%s/meta/manifest' % (self.url,
-                                                    _get_path(entity_id))
+            manifest_url = '{}/{}/meta/manifest'.format(self.url,
```
Comment (frankban) [inline theblues/charmstore.py:227]: Again here, yeah we should have an `add_channel(url, channel=None) -> url` function somewhere.

**Comment by frankban** [comment]
:+1: with avoiding repeating the add channel logic everywhere.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/theblues-lib/68/
Test PASSed.

File: theblues/charmstore.py (hunk 1 above)
Comment (fabricematrat) [inline theblues/charmstore.py]: Done.

File: theblues/charmstore.py (hunk 2 above)
Comment (fabricematrat) [inline theblues/charmstore.py]: Done.

File: theblues/charmstore.py (hunk 3 above)
Comment (fabricematrat) [inline theblues/charmstore.py:227]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/theblues-lib/69/
Test PASSed.

**Comment by urosj** [comment]
:+1: did not do QA

**Comment by fabricematrat** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/theblues-lib-merge

**Pull request merged**